docs/client-specifications/configuration.md

```diff
@@ -0,0 +1,147 @@
+Documentation must clearly indicate that these values may change over time if Momento finds in the future that some
+different default values will provide a better or more performant user experience.
+
+All configuration objects must also provide a `default()` function that returns the original "blessed" defaults from
+Momento.  Documentation must clearly state that these values will not change over time; users can rely on them to
+be stable across new minor version releases of the SDKs.
+
+The `SimpleCacheClient` constructor should expose a required argument that is an instance of
+
+## Required pre-built configuration objects
+
+Each SDK must provide at least the following out-of-the-box configuration objects:
+
+```typescript
+import * as config from '@gomomento/sdk/config';
+
+/**
+ * Dev provides defaults suitable for a medium-to-high-latency dev environment.  Permissive timeouts, retries, potentially
+ * a higher number of connections, etc.
+ */
+const devConfig = config.Dev.latest();
+
+/**
+ * ProdLowLatency provides defaults suitable for a low-latency prod environment.  This is for the use case where the
+ * client is running on an application server in the same region as the Momento service.  It has more agressive timeouts
+ * and retry behavior, to ensure that cache unavailability doesn't force unacceptably high application latencies.
+ */
+const prodConfig = config.ProdLowLatency.latest();
+```
+
+## Configuration interface
+
+The configuration interface provides nested objects for retry strategy, middleware, and transport:
+
+```typescript
+interface Configuration {
+  retryStrategy: RetryStrategy,
+  middleware: Middleware,
+  transportStrategy: TransportStrategy
+}
+```
+
+### RetryStrategy
+
+The `RetryStrategy` provides a function that is called when a request fails.  This function is responsible for determining
+whether the request should be retried, and if so, how long the client should delay before retrying:
+
+```typescript
+interface RetryStrategy {
+  /**
+   * Returns the number of milliseconds after which the request should be retried, or `undefined` if the request should
+   * not be retried.  `0` means retry immediately.
+   */
+  determineWhenToRetryRequest(grpcResult: GrpcResult, grpcRequest: GrpcRequest, attemptNumber: number): Promise<number | undefined>;
+}
+```
+
+The simplest default implementation of this strategy can expose a `maxAttempts` integer, and always return `0` for
+immediate retries until the maximum number of attempts is reached.  A slightly more sophisticated initial implementation
+would have basic support for exponential backoff and jitter.
+
+### Middleware
+
+The `Middleware` interface allows the Configuration to provide a higher-order function that wraps all requests.  This
+allows future support for things like client-side metrics or other diagnostics helpers.
+
+```typescript
+interface Middleware {
+  wrapRequest(middlewareFn: (GrpcRequest) => GrpcResult): GrpcResult;
```

Comment:
Should the signature for our middleware be protocol agnostic? Should we be able to switch out GRPC without the user's middleware breaking? 

Just want to call out since before we had been trying to hide GRPC from any client-facing exposed bits in past.

Reply:
This is a great question.

I would expect that most users will never implement this, they will just use our pre-builts.

For the advanced users that do want to implement this, I... think it's probably fair game to say that they have opted in to that complexity?  For the middleware to be maximally useful I think we would have to provide a wrapper data class that included all of the gRPC info (metadata etc.) anyway, so I'm not sure it's worth the abstraction?  But I could be convinced to reconsider.

Reply:
I think this is ok and safest place to start. We can always revisit in v2 version if we get alot of feedback. Would rather avoid abstracting to much upfront.

Reply:
agree